<div style="direction:rtl;line-height:300%;">
<font face="XB Zar" size=5>
<div align=center>
<font face="B Titr" size=5>
<p></p><p></p>
بسمه تعالی
<p></p>
</font>
<p></p>
<font>
<br>
درس بازیابی پیشرفته اطلاعات
<br>
مدرس: دکتر سلیمانی
</font>
<p></p>
<br>
<font>
<b>فاز اول پروژه</b>
</font>
<br>
<br>
موعد تحویل: ۲۸ اسفند ۱۳۹۹
<br>
<br>
<font size=4.8>
دستیاران آموزشی: نیما جمالی، آرمین سعادت، ایمان غلامی
</font>
<br>
<br>
<br>
<font>
دانشگاه صنعتی شریف
<br>
دانشکده مهندسی کامپیوتر
<br>
<br>
</font>
</div>
</font>
</div>

<p></p>
<div dir="rtl">
<font face="XB Zar" size=4>
    <h1>
    <b>مقدمه</b>
    </h1>
    <p></p>
    <p></p>
    هدف از فاز اول پروژه، طراحی و پیاده‌سازی سیستم بازیابی اطلاعات برای مجموعه دادگان تعیین شده می‌باشد.<br>
    اولین مرحله، پیش‌پردازش مجموعه دادگان است. پس از آن نمایه‌ها با ویژگی‌های خواسته شده پیاده‌سازی می‌شود. در گام بعدی به ذخیره و بازخوانی نمایه‌ها به همراه روش‌های فشرده‌سازی پرداخته می‌شود. پس از آن تکنیک‌های اصلاح پرسمان پیاده‌سازی شده و در نهایت جستجو روی دادگان صورت می‌گیرد. همچنین در بخش آخر با پیاده‌سازی برخی معیارهای ارزیابی، عملکرد سیستم مورد سنجش قرار می‌گیرد.<br><br>
     توضیحات مربوط به هر بخش در ادامه آمده است که اهداف، محدودیت‌ها و خواسته‌های آن بخش را مشخص می‌کند.
     در هر بخش توابعی مشخص شده است که محتوای آن با کد نوشته شده توسط شما باید پر شود. شما می‌توانید در همین فایل، سیستم بازیابی خود را پیاده‌سازی کنید یا فایل‌های خودتان را ایمپورت کرده و از آن‌ها استفاده کنید. در هر صورت تمامی کدهای لازم را به همراه این نوت‌بوک ارسال کنید. همچنین در نظر داشته باشید که ملاک اصلی نمره‌دهی شما اجرای صحیح توابع مشخص‌شده در این نوت‌بوک می‌باشد. بنابراین از صحت اجرای نوت‌بوک خود اطمینان پیدا کنید.<br><br>
     تنها زبان قابل قبول برای پروژه پایتون است. محدودیت استفاده از کتاب‌خانه‌های آماده در هر بخش مشخص شده است. پروژه ۱۱۵ نمره دارد که ۱۵ نمره از آن امتیازی و مربوط به بخش اصلاح پرسمان می‌باشد.
    
    
</font>
</div>
<p></p>

<div dir="rtl">
<font face="XB Zar" size=4>
    <h1>
    <b>مجموعه دادگان</b>
    </h1>
    مجموعه دادگان مورد بررسی در این پروژه از سایت kaggle فراهم شده است. این مجموعه شامل اطلاعات ۶۰۰۰ فیلم سینمایی از سال ۱۹۰۴ تا ۲۰۱۷ است. داده‌ها در قالب فایل csv   دارای ستون‌‌های plot، title، id می‌باشد. id یک شناسه یکتا برای هر فیلم است که برای ارزیابی بهتر عملکرد شما به داده‌ها اضافه شده است و در مجموعه دادگان اصلی وجود نداشته است. همانطور که می‌دانید برای پیاده‌سازی نمایه باید به هر داکیومنت یک شناسه اختصاص بدهید. <b>شناسه مربوط به هر داکیومنت باید id ذکر شده برای آن در مجموعه دادگان باشد.</b> هر فیلم از دو بخش title و plot  تشکیل شده است که از این دو بخش در ساخت نمایه و جستجو استفاده می‌شود. plot خلاصه‌ای از طرح داستان فیلم است.<br>
    علاوه بر مجموعه دادگان اصلی، تعدای پرسمان در اختیار شما قرار گرفته است. همچنین جواب مطلوب هر یک از این پرسمان‌ها نیز فراهم شده که طبیعتا زیرمجموعه‌ای از مجموعه دادگان است. شما باید از این پرسمان‌ها و نتیجه مورد انتظار هر کدام برای ارزیابی سامانه خود استفاده کنید.
    پرسمان‌ها و شناسه فیلم‌های بازیابی شده مورد انتظار از هر پرسمان در فایل validation.json آمده است. توضیحات بیشتر در رابطه با استفاده از این فایل در بخش ارزیابی آمده است.
</font>
</div>

<div dir="rtl">
<font face="XB Zar" size=4>
    <h1>
    <b>پیش‌پردازش و آماده‌سازی داده‌ها (۱۵ نمره)</b>
    </h1>
    در این بخش ابتدا داده‌ها را از فایل بخوانید. برای آماده‌سازی متن می‌توانید از کتاب‌خانه‌های آماده استفاده کنید. یکی از کتاب‌خانه‌های معروف برای این کار <a href="https://www.nltk.org/">NLTK</a> است اما در انتخاب روش پیاده‌سازی این بخش مختارید. برای این بخش باید تابع ()prepare_text را تکمیل کنید. این تابع یک متن انگلیسی ورودی گرفته و توکن‌‌های مربوط به آن‌را در قالب یک لیست خروجی می‌دهد. متن ورودی در عمل تایتل یا طرح داستان هر فیلم است. دقت کنید که لیست خروجی شامل تعدادی توکن است که عملیات case folding، stemming و lemmatization روی آن‌ها اجرا شده است. در ضمن علائم نگارشی نباید به عنوان توکن در نظر گرفته شود. در کد زیر یک نمونه ورودی و خروجی نمایش داده شده است. با توجه به نحوه پیاده‌سازی انواع بازگردانی به ریشه قابل قبول است.
</font>
</div>

In [1]:
# TAVVAJJOH: agar wordnet ra nadarid avval 2 khatte zir ra run konid (be vpn niaz dare baraye download)

# import nltk
# nltk.download('wordnet')

In [89]:
import nltk
import string
from nltk.stem.porter import PorterStemmer
from nltk.stem import WordNetLemmatizer
import csv

# man az in link baraye tabe komak gereftam: https://tomelf.github.io/nlp/machine%20learning/text-preprocessing/
# va az in link baraye khoondane file .csv: https://realpython.com/python-csv/

lemmatizer = WordNetLemmatizer()
stemmer = PorterStemmer()

all_original_words={}

def tokenize(raw_text, add_to_words=True):
    global all_original_words
    raw_text = raw_text.lower()
    my_dict={}
    for p in list(string.punctuation):
        my_dict[p]=' '
    my_dict['\n']=' '
    my_dict["'"]=''
    
    trantab = str.maketrans(my_dict)
    raw_text = raw_text.translate(trantab)
    
    tokens=nltk.word_tokenize(raw_text)
    if add_to_words:
        for t in tokens:
            all_original_words[t]=True
        
    return tokens

def prepare_text(raw_text, add_to_words=True):
    """Preprocesses the text with tokenization, case folding, stemming and lemmatization

    Parameters
    ----------
    raw_text : str
        The title or plot of a movie

    Returns
    -------
    list
        A list of tokens
    """
    global lemmatizer,stemmer
    
    tokens = tokenize(raw_text, add_to_words)
    tokens = [lemmatizer.lemmatize(lemmatizer.lemmatize(t,'v'),'a') for t in tokens]
    tokens = [stemmer.stem(d) for d in tokens]
    
    return tokens
    # TODO: tokenize, case_folding, stem, lemmatize
    # return ["edvard", "am", "a", "run", "he", "am", "always", "run"]

prepare_text("Edvard was a Runner. He was always running.")


movies_file=[]

def read_csv_file(file_name):
    ans=[]
    csv_file=open(file_name,'r')
    csv_reader = csv.reader(csv_file, delimiter=',')
    flag = False
    for row in csv_reader:
        if flag:
            ans.append([int(row[0]),prepare_text(row[1]),prepare_text(row[2])])
        flag = True
    return ans

def main():
    global movies_file
    movies_file=read_csv_file('movies.csv')
    
    
main()

<div dir="rtl">
<font face="XB Zar" size=4>
    <h1>
    <b>شناسایی و حذف stop-words (۵ نمره)</b>
    </h1>
    این بخش باید توسط خودتان و بدون استفاده از کد آماده پیاده‌سازی شود. ترم‌های موجود در مجموعه دادگان را بر اساس تکرار آن‌ها مرتب کرده و پرتکرارترین آن‌ها را به عنوان stop-words در نظر بگیرید. اینکه چند ترم را به عنوان stop-words در نظر بگیرید به عهده خودتان است.<br>
    با فراخوانی تابع ()get_stop_words لیست stop-words به همراه تعداد تکرار آن‌‌ها خروجی داده می‌شود.<br>
    ترم‌های به دست آمده از این بخش نباید در نمایه حضور داشته باشند.
</font>
</div>

In [6]:
NUMBER_OF_STOP_WORDS = 25
# punctuations='!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'
movies_without_sw=[]
sw=[]
def drop_stop_words(tokens):
    global sw
    out=[]
    for t in tokens:
        if t not in sw:
            out.append(t)
    return out

def get_stop_words():
    """Detects stop-words

    Returns
    -------
    dict
        A dictionary of terms with their repetition
    """
    global NUMBER_OF_STOP_WORDS,movies_file,movies_without_sw,sw
    
    cnt = {}
    '''my_dict={}
    for p in list(punctuations):
        my_dict[p]=' '
    my_dict['\n']=' '
    my_dict["'"]=''
    trantab = str.maketrans(my_dict)
    
    f = open("movies.csv", "r")
    tokens = f.read().lower().translate(trantab).split(' ')
    for t in tokens:
        if ' ' not in t and t is not "":
            if t in cnt:
                cnt[t] += 1
            else:
                cnt[t] = 0'''
    for row in movies_file:
        for t in row[1]+row[2]:
            if t in cnt:
                cnt[t] += 1
            else:
                cnt[t] = 0
    out={}
    sw2=[]
    
    for token in sorted(cnt, key=cnt.get, reverse=True)[:NUMBER_OF_STOP_WORDS]:
        out[token]= cnt[token]
        sw2.append(token)
    
    if sw!=sw2:
        sw=sw2
        movies_without_sw=[]
        for row in movies_file:
            movies_without_sw.append([row[0],drop_stop_words(row[1]),drop_stop_words(row[2])])
    #print(movies_without_sw[:5])
    return out
    
    # TODO: find top most repetitive terms and report them
    #return {"a":1000, "the":800, "am":500}

get_stop_words()

{'the': 147457,
 'to': 91727,
 'and': 83264,
 'be': 69128,
 'a': 68047,
 'of': 42860,
 'in': 38981,
 'hi': 34775,
 'he': 32197,
 'her': 28016,
 'that': 26294,
 'with': 25062,
 'have': 20202,
 'him': 18921,
 'by': 17854,
 'for': 17607,
 'she': 17514,
 'as': 15645,
 'on': 14960,
 'but': 13529,
 'they': 13157,
 'who': 12731,
 'at': 12352,
 'from': 11097,
 'an': 10858}

<div dir="rtl">
<font face="XB Zar" size=4>
    <h1>
    <b>نمایه‌سازی (۱۵ نمره)</b>
    </h1>
    در این بخش باید برای سامانه positional index بسازید. برای هر ترم باید مشخص باشد که آن ترم در تایتل چه فیلم‌هایی و در چه جایگاهی از تایتل هر فیلم قرار گرفته است. همچنین برای هر ترم باید مشخص باشد که آن ترم در طرح داستان چه فیلم‌هایی و در چه جایگاهی از طرح داستان هر فیلم قرار گرفته است.<br>
     برای ارزیابی بهتر و عادلانه‌تر به ویژه برای ارزیابی بخش فشرده‌سازی، از استاندارد بیان شده در قطعه کد زیر برای ذخیره posting list هر ترم در RAM استفاده کنید.
</font>
</div>

In [101]:
positional_index={}

def add_row_2_positional_index(row):
    global positional_index
    j=row[0]
    for i in range(len(row[1])):
        term=row[1][i]
        lst=positional_index[term][0]
        if len(lst)==0 or lst[-1][0]!=j:
            lst.append([j,[i]])
        else:
            lst[-1][1].append(i)

    for i in range(len(row[2])):
        term=row[2][i]
        lst=positional_index[term][1]
        if len(lst)==0 or lst[-1][0]!=j:
            lst.append([j,[i]])
        else:
            lst[-1][1].append(i)
    
def create_positional_index():
    """Creates positional index for both title and plot
    
    positional_index = {'term': [title_posting_list, plot_posting_list]}
    posting_list = [[doc_1_id, [doc_1_occurrance_index_1, ...]], [doc_2_id, [doc_2_occurrance_index_1, ...]], ...]
    doc_ids and doc_occurrence_indexes are sorted


    Returns
    -------
    dict
         The created positional_index
    """
    
    '''title_posting_list = [[12, [10]], [123, [5, 6]]]
    plot_posting_list = [[1, [1]], [3, [50, 60]]]
    positional_index = {'term': [title_posting_list, plot_posting_list]}'''
    
    global movies_without_sw, positional_index
    
    positional_index = {}
    all_terms={}
    
    for row in movies_without_sw:
        for token in row[1]+row[2]:
            all_terms[token]=True
    all_terms=list(all_terms)
    
    
    
    #print(len(all_terms),all_terms[:10])
    
    for term in all_terms:
        title_pl=[]
        plot_pl=[]
        positional_index[term]=[[],[]]
    
    for jj in range(len(movies_without_sw)):
        row=movies_without_sw[jj]
        add_row_2_positional_index(row)
            

    return positional_index    
create_positional_index()['sherlock']

[[[39, [0]],
  [854, [1]],
  [1138, [0]],
  [1224, [0]],
  [1225, [0]],
  [1226, [0]],
  [1392, [0]],
  [5597, [0]]],
 [[762, [135]],
  [858, [0]],
  [1138, [163]],
  [1224, [0]],
  [1225, [67, 77]],
  [1335, [2]],
  [3235, [12]],
  [3417, [9, 38]],
  [5597, [3]]]]

<div dir="rtl">
<font face="XB Zar" size=4>
    <h1>
    <b>پویا‌سازی نمایه (۱۰ نمره)</b>
    </h1>
    نمایه ایجاد شده باید قابلیت حذف و اضافه تک داکیومنت را داشته باشد.
    برای اضافه شدن داکیومنت، به تابع ()add_single_document یک رشته داده می‌شود که اطلاعات مربوط به داکیومنت شامل id و plot و title در آن با کاما جدا شده است. برای حذف داکیومنت نیز id آن به تابع ()remove_single_document داده می‌شود.<br>
    تضمین می‌شود که شرط یکتا بودن id داکیومنت‌ها نقض نشود. برای مثال دو داکیومنت با شناسه یکسان به مجموعه اضافه نخواهد شد. البته ممکن است حذف شده و دوباره اضافه شود.
</font>
</div>

In [102]:
def add_single_documnet(document):
    """Adds a document to positional index

    Parameters
    ----------
    document : str
        Comma separated string containing id,title,plot in this exact order
    """

    # TODO: implement adding a document to the index
    inp=document.split(',')
    inp=inp[0:3]
    inp[0]=int(inp[0])
    for i in range(1,3):
        inp[i]=drop_stop_words(prepare_text(inp[i]))
        #print(inp[i])
    add_row_2_positional_index(inp)
    return

new_document = "1234,The Adventures of Sherlock Holmes,The picture begins with Moriarty and Holmes verbally sparring on the steps"
add_single_documnet(new_document)

In [103]:
def remove_signle_document(document_id):
    """removes a document from positional index

    Parameters
    ----------
    document_id : int
        Id of the document
    """
    
    # TODO: implement removing a document from the index
    document_id=int(document_id)
    for token in positional_index:
        pre_lst=positional_index[token]
        new_lst=[[],[]]
        for i in range(2):
            for lst in pre_lst[i]:
                if lst[0]!=document_id:
                    new_lst[i].append(lst)
        positional_index[token]=new_lst
    return
  
remove_signle_document(1234)

<div dir="rtl">
<font face="XB Zar" size=4>
    <h1>
    <b>ذخیره و فشرده‌سازی نمایه (۲۰ نمره)</b>
    </h1>
    در این بخش باید توانایی ذخیره کردن نمایه و بارگذاری مجدد آن را به سامانه اضافه کنید. ذخیره‌سازی به ۳ روش صورت می‌گیرد. بدون فشرده‌سازی، فشرده‌سازی از روش gamma-code و فشرده‌سازی از روش variable-byte. روش‌های فشرده‌سازی باید توسط خودتان پیاده‌سازی شود.
    برای ذخیره نمایه در فایل نیز از JSON  استفاده کنید.<br>
     بخشی از نمره شما در این قسمت به میزان فشرده‌سازی نمایه اختصاص داده شده است. بنابراین پیاده‌سازی بهینه روش‌های فشرده‌سازی مهم است.
</font>
</div>

In [95]:
import json
from pathlib import Path
from copy import deepcopy

base=35

def chr2(x):
    if x==ord('\\'):
        x=125
    return chr(x)

def variable_len_enc(x):
    ans=0
    cnt=0
    while x>0 or cnt==0:
        ans<<=1
        ans|=1
        ans<<=7
        ans|=x&((1<<7)-1)
        #print(ans,x,cnt)
        x>>=7
        cnt+=1
    ans^=1<<7
    ans2=0
    while ans>0:
        ans2<<=8
        ans2|=ans&((1<<8)-1)
        ans>>=8
        
    return ans2,cnt

def prepare_list_enc(plist):
    lst=[plist[0]]+plist[1]
    for i in range(len(lst)-1,1,-1):
        if lst[i]>lst[i-1]:
            lst[i]-=lst[i-1]
        else:
            print('Error in variable len enc function!')
    return lst

def variable_len_enc_list(plist):
    global base
    lst=prepare_list_enc(plist)
    
    ans=0
    cnt=0
    out=""
    for x in lst:
        enc=variable_len_enc(x)
        cnt+=8*enc[1]
        ans<<=8*enc[1]
        ans|=enc[0]
        while cnt>=6:
            cnt-=6
            out+=chr2(base+(ans>>cnt))
            ans&=(1<<cnt)-1
    if cnt>0:
        out+=chr2(base+(ans<<(6-cnt)))
    
    return out
    # har 6 bit ra daroon yek char magozaram! chon tedadi az char ha
    # non-printable hastan va be tor khoobi dar json zakhire nemishavand!

def gamma_enc(x):
    b=bin(x)
    s='1'*(len(b)-3)+'0'+b[3:]
    return int(s,2),len(s)



def gamma_enc_list(plist):
    global base
    lst=prepare_list_enc(plist)
    
    ans=0
    cnt=0
    out=""
    for x in lst:
        enc=gamma_enc(x+1)
        cnt+=enc[1]
        ans<<=enc[1]
        ans|=enc[0]
        while cnt>=6:
            cnt-=6
            out+=chr2(base+(ans>>cnt))
            ans&=(1<<cnt)-1
    if cnt>0:
        out+=chr2(base+(ans<<(6-cnt))+(1<<(6-cnt))-1)
    
    return out
    # har 6 bit ra daroon yek char magozaram! chon tedadi az char ha
    # non-printable hastan va be tor khoobi dar json zakhire nemishavand!
    
    

def store_index(path, compression_type):
    """Stores the index in a file

    Parameters
    ----------
    path : str
        Path to store the file

    compression_type : str
        Could be one of the followings:
        - no-compression
        - gamma-code
        - variable-byte

    Returns
    -------
    int
        The size of the stored file
    """
    global positional_index
    try:
        f = open(path, "x")
    except:
        pass
    f=open(path,'w')

    if compression_type=='no-compression':
        f.write(json.dumps(positional_index))
        
    elif compression_type=='variable-byte':
        new_positional_index={}
        for token in positional_index:
            new_positional_index[token]=[[],[]]
            for i in range(2):
                for lst in positional_index[token][i]:
                    new_positional_index[token][i].append(variable_len_enc_list(lst))
            
        f.write(json.dumps(new_positional_index))
        
    elif compression_type=='gamma-code':
        new_positional_index={}
        for token in positional_index:
            new_positional_index[token]=[[],[]]
            for i in range(2):
                for lst in positional_index[token][i]:
                    new_positional_index[token][i].append(gamma_enc_list(lst))
            
        f.write(json.dumps(new_positional_index))
    
    size_of_file = 1024
    
    f.close()
    
    return Path(path).stat().st_size

#store_index("path/to/file.json", 'gamma-code')

12213935

In [64]:
#print(gamma_dec_list(gamma_enc_list([0,[824,829,215406,33333333,99999999889999]])))

[0, [824, 829, 215406, 33333333, 99999999889999]]


In [97]:
import json 

base=35

def ord2(c):
    if ord(c)==125:
        return ord('\\')
    return ord(c)

def ok_lst(lst, offset):
    if len(lst)<2:
        print('Error in ok_lst!', lst)
        
    for i in range(len(lst)):
        lst[i]+=offset
        
    ans=[lst[1]]
    for x in lst[2:]:
        ans.append(ans[-1]+x)
    return [lst[0],ans]

def variable_len_dec_list(s):
    global base
    ans=0
    cnt=0
    lst=[]
    for c in s:
        x=ord2(c)-base
        cnt+=6
        ans<<=6
        ans|=x
        while cnt>=8:
            cnt-=8
            z=ans>>cnt
            if (z&(1<<7))==0:
                lst.append(0)
            lst[-1]<<=7
            lst[-1]|=z&((1<<7)-1)
            ans&=(1<<cnt)-1
    
    return ok_lst(lst,0)

def gamma_dec_list(s):
    global base
    ans=0
    cnt=0
    lst=[]
    mod=0
    
    for c in s:
        x=ord2(c)-base
        x=bin(x)[2:]
        x='0'*(6-len(x))+x
        #print(x)
        i=0
        while i<len(x):
            if mod==0:
                while i<len(x):
                    cnt+=1
                    i+=1
                    if x[i-1]=='0':
                        mod=1
                        ans=1
                        cnt-=1
                        break
            if mod==1:
                while cnt>0 and i<len(x):
                    cnt-=1
                    ans<<=1
                    if x[i]=='1':
                        ans|=1
                    i+=1

                if cnt==0:
                    lst.append(ans)
                    mod=0
    
    return ok_lst(lst,-1)

def load_index(path, compression_type):
    """Loads the index from a file

    Parameters
    ----------
    path : str
        Path of the file to load from

    compression_type : str
        Could be one of the followings:
        - no-compression
        - gamma-code
        - variable-byte
    """
    global positional_index
    # TODO: load and decompress positional_index
    f=open(path,'r')
    if compression_type=='no-compression':
        positional_index2=json.loads(f.read())
    elif compression_type=='variable-byte':
        positional_index2=json.loads(f.read())
        for token in positional_index2:
            for i in range(2):
                lst=[]
                for s in positional_index2[token][i]:
                    lst.append(variable_len_dec_list(s))
                positional_index2[token][i]=lst
    elif compression_type=='gamma-code':
        positional_index2=json.loads(f.read())
        for token in positional_index2:
            for i in range(2):
                lst=[]
                for s in positional_index2[token][i]:
                    lst.append(gamma_dec_list(s))
                positional_index2[token][i]=lst
    
    
    #print(positional_index==positional_index2)
    positional_index=positional_index2
    f.close()
    return

#load_index("path/to/file.json", 'gamma-code')

<div dir="rtl">
<font face="XB Zar" size=4>
    <h1>
    <b>اصلاح پرسمان (۱۵ نمره امتیازی)</b>
    </h1>
    در صورتی که پرسمان ورودی دارای غلط املایی باشد یا به عبارتی لغاتی از آن در لغت‌نامه موجود نباشد، لازم است که با جستجوی لغت‌های احتمالی و انتخاب بهترین لغت به ادامه‌ی جستجو با پرسمان اصلاح شده پرداخته شود. برای اینکار ابتدا باید با روش bigram و معیار jaccard نزدیک‌ترین لغات به لغت با غلط املایی را پیدا کنید. سپس با استفاده از معیار edit distance بهترین لغت را از میان آن‌ها بیابید.<br>
    نیازی به ذخیره‌سازی و فشرده‌سازی نمایه بایگرم نیست. همچنین می‌توانید از کد آماده برای محاسبه edit distance استفاده کنید.
</font>
</div>

In [99]:
# baraye edit distance az site 'geeks for geeks' estefade kardam

def editDistDP(str1, str2):
    m=len(str1)
    n=len(str2)
    # Create a table to store results of subproblems
    dp = [[0 for x in range(n + 1)] for x in range(m + 1)]
 
    # Fill d[][] in bottom up manner
    for i in range(m + 1):
        for j in range(n + 1):
 
            # If first string is empty, only option is to
            # insert all characters of second string
            if i == 0:
                dp[i][j] = j    # Min. operations = j
 
            # If second string is empty, only option is to
            # remove all characters of second string
            elif j == 0:
                dp[i][j] = i    # Min. operations = i
 
            # If last characters are same, ignore last char
            # and recur for remaining string
            elif str1[i-1] == str2[j-1]:
                dp[i][j] = dp[i-1][j-1]
 
            # If last character are different, consider all
            # possibilities and find minimum
            else:
                dp[i][j] = 1 + min(dp[i][j-1],        # Insert
                                   dp[i-1][j],        # Remove
                                   dp[i-1][j-1])    # Replace
 
    return dp[m][n]

bigram={}
def get_biwords(word):
    word='$'+word+'$'
    biwords=sorted([word[i]+word[i+1] for i in range(len(word)-1)])
    out=[]
    for x in biwords:
        if len(out)==0 or out[-1]!=x:
            out.append(x)
    return out

def create_bigram_index():
    """Creates the bigram index for spell correction"""
    global all_original_words,bigram
    
    
    bigram = {}
    
    lst=sorted(all_original_words)
    all_original_words={}
    for x in lst:
        all_original_words[x]=True
        
    all_chars=['$']+[chr(x) for x in range(ord('a'),ord('z')+1)]
    all_pairs=[x+y for x in all_chars for y in all_chars]
    
    
    for x in all_original_words:
        flag=False
        for c in x:
            if ord(c)<ord('a') or ord(c)>ord('z'):
                flag=True
        if flag:
            continue
        for biword in get_biwords(x):
            if biword not in bigram:
                bigram[biword]=[]
            bigram[biword].append(x)
    #print(cnt/len(all_original_words))
    return bigram

def create_bigram_main():
    create_bigram_index()
create_bigram_main()

In [100]:
def jaccard_value(x,y):
    disjoint=0
    bx=get_biwords(x)
    by=get_biwords(y)
    for t in bx:
        if t in by:
            disjoint+=1
    union=len(bx)+len(by)-disjoint
    if union==0:
        return 0.0
    return disjoint/union
    
def get_corrected_word(token):
    global all_original_words,bigram
    NUMBER_OF_BEST_JACC=5
    jacc_value={}
    
    if token in all_original_words:
        return token
    
    for b in get_biwords(token):
        if b not in bigram:
            continue
        for t in bigram[b]:
            if t in jacc_value:
                continue
            jacc_value[t]=jaccard_value(token,t)
    
    top_jaccard=[x[0] for x in sorted(jacc_value.items(),key=lambda x:x[1])[-NUMBER_OF_BEST_JACC:]][::-1]
    
    if len(top_jaccard)==0:
        return token
    best=top_jaccard[0]
    best_edit_dist=editDistDP(token,best)
    
    for x in top_jaccard:
        tmp=editDistDP(token,x)
        #print(x,tmp)
        if tmp<best_edit_dist:
            best=x
            best_edit_dist=tmp
    return best

def get_corrected_text(raw_text):
    """Corrects the query

    Parameters
    ----------
    raw_text : str
        Input text that could be a title or a plot

    Returns
    -------
    str
        The corrected text
    """

    # TODO: correct the query using bigram, jaccard, and edit-distance
    tokens=tokenize(raw_text,False)
    #print(tokens)
    ans=[]
    for t in tokens:
        ans.append(get_corrected_word(t))
        
    return " ".join(ans)

get_corrected_text("the adevntures of herlock holmes")

'the adventures of sherlock holmes'

<div dir="rtl">
<font face="XB Zar" size=4>
    <h1>
    <b>جستجو و بازیابی اسناد (۲۵ نمره)</b>
    </h1>
    در این بخش جستجو ترتیب‌دار در فضای برداری tf-idf به روش <b>lnn-ltn</b> انجام می‌شود. یک پرسمان title و یک پرسمان plot ورودی گرفته شده و هر کدام در بخش مربوطه از اسناد جستجو می‌شوند. امتیاز نهایی هر سند برابر با جمع وزن‌دار امتیاز به دست آمده از جستجو در بخش title و plot است. به این صورت که وزن plot واحد در نظر گرفته شده و وزن تایتل به عنوان ورودی داده می‌شود.<br>
    در نهایت اسناد برتر را نمایش دهید. تعداد حداکثر اسناد برتر نیز به عنوان ورودی داده می‌شود.<br><br>
    نکته بسیار مهم نحوه نمایش اسناد انتخابی است. برای نمایش هر سند علاوه بر استفاده از شناسه و تیتر، یک هایلایت برای آن درست کنید. به این معنا که کلمات موجود در پرسمان را که باعث انتخاب سند شده‌اند به همراه ۲-۳ ترم قبل و بعد از آن به عنوان هایلایت آن سند نمایش دهید. اینگونه کاربر می‌تواند خیلی سریع دلیل بازیابی اسناد توسط سامانه را متوجه شود. مشابه کاری که سرچ گوگل انجام می‌دهد و ۲-۳ خط مربوطه را زیر وبسایت‌های پیشنهادی نمایش می‌دهد. طبیعتا راه حل بهینه برای این‌کار استفاده از قابلیت‌های نمایه جایگاهی می‌باشد.<br>
</font>
</div>

In [93]:
from math import log

def single_query(query,i):
    global positional_index,movies_without_sw
    
    sample_token={}
    query=get_corrected_text(query)
    query=prepare_text(query, False)
    N=len(movies_without_sw)
    score_doc={}
    query_unique=[]
    for t in sorted(query):
        if len(query_unique)==0 or query_unique[-1]!=t:
            query_unique.append([t,1])
        elif len(query_unique)!=0:
            query_unique[-1][1]+=1
    
    for t,cnt in query_unique:
        if t not in positional_index:
            continue
        df=len(positional_index[t][i])
        for lst in positional_index[t][i]:
            d=lst[0]
            w=(1+log(len(lst[1]),10))*log(N/df,10)*(1+log(cnt,10))
            if d not in score_doc:
                score_doc[d]=0
            score_doc[d]+=w
            sample_token[d]=t
    return score_doc,sample_token

def search(title_query, plot_query, title_weight, max_result_count=10):
    """Finds relevant documents to query

    Parameters
    ----------
    title_query : str
        Query to search in titles

    plot_query : str
        Query to search in plots

    title_weight : int
        Defines the significance of title

    max_result_count : int
        Max number of retrieved documents
    
    Returns
    -------
    list
        Retreived documents with snippet
    """

    # TODO: implement lnn-ltn search and show results with highlight
    '''doc_id = 1
    title = "Call <b>wild</b>"
    highlight = "<b>florence</b> lawrence already most <b>popular</b> among biograph "
    return [
            [doc_id, title, highlight],
            [doc_id, title, highlight],
            [doc_id, title, highlight],
    ]
    idf[t]=log(10) N/df[t]
    df[t]=len(positinoal_index[t][i])
    tf[t,d]=#t in d 
    '''
    global positional_index,movies_without_sw
    
    id2index={}
    for i in range(len(movies_without_sw)):
        row=movies_without_sw[i]
        id2index[row[0]]=i
    
    title_score=single_query(title_query,0)[0]
    plot_score,sample_token=single_query(plot_query,1)
    score={}
    
    for d in title_score:
        if d not in plot_score:
            plot_score[d]=0
    
    for d in plot_score:
        if d not in title_score:
            title_score[d]=0
        score[d]=plot_score[d]+title_weight*title_score[d]
    
    result=[]
    for d in sorted(score, key=score.get, reverse=True)[:max_result_count]:
        tit=""
        lst=movies_without_sw[id2index[d]][1]
        for t in lst:
            if t in title_query:
                tit+="<b>"+t+"</b> "
            else:
                tit+=t+" "
        tit=tit[:-1]
        lst=movies_without_sw[id2index[d]][2]
        highlight=""
        pos=0
        if d in sample_token:
            token=sample_token[d]
            for x in positional_index[token][1]:
                if x[0]==d:
                    pos=max(0,x[1][0]-3)
                    break
        
        for t in lst[pos:pos+7]:
            if t in plot_query:
                highlight+="<b>"+t+"</b> "
            else:
                highlight+=t+" "
        highlight=highlight[:-1]
        result.append([d,tit,highlight])
    return result

search("wild", "Florence popular", 10, 3)

[[4,
  'call <b>wild</b>',
  'lawrenc alreadi most <b>popular</b> among biograph compani'],
 [4864,
  '<b>wild</b> thing',
  '<b>popular</b> miami area high school guidanc counselor'],
 [2238,
  'davi crockett king <b>wild</b> frontier',
  'elect crockett becom <b>popular</b> member hous repres']]

<div dir="rtl">
<font face="XB Zar" size=4>
    <h1>
    <b>ارزیابی عملکرد سامانه (۱۰ نمره)</b>
    </h1>
    همانطور که در بخش مربوط به مجموعه دادگان گفته شد، تعدادی پرسمان نمونه به همراه اسناد مورد نظر برای آن‌ها در اختیار شما قرار گرفته است. در هر مورد اطلاعات لازم برای ایجاد یک پرسمان ذکر شده است. مطابق آن‌ها پرسمان خود را ایجاد کنید. نتایج به دست آمده از هر پرسمان را به عنوان predicted results آن پرسمان در نظر بگیرید. همچنین در هر مورد لیستی از شناسه‌ها وجود دارد. این لیست را به عنوان actual results در نظر بگیرید. <br>
    معیار‌های زیر را پیاده‌سازی کنید (بدون استفاده از کد آماده) و نتیجه این معیارها را روی مجموعه‌ی actual و predicted گزارش کنید. دقت کنید که به ازای هر پرسمان باید تمام معیارها را در قالب یک جدول گزارش کنید.<br> دقت کنید که ۳ پرسمان آخر فقط برای افرادی که بخش spell correction را پیاده کرد‌ه‌اند نتیجه مطلوب دارد.
</font>
</div>

In [21]:
def tpfnfp(actual, predicted):
    ans=0
    for x in predicted:
        if x in actual:
            ans+=1
            
    return ans,len(actual)-ans,len(predicted)-ans


def precision(actual, predicted):
    """Calculate precision

    Parameters
    ----------
    actual : list
        [1st_doc_id, 2nd_doc_id, ...]

    predicted : list
        [1st_doc_id, 2nd_doc_id, ...]

    Returns
    -------
    float
        Precision score
    """

    # TODO: implement precision
    tmp = tpfnfp(actual,predicted)
    return tmp[0]/(tmp[0]+tmp[2])

precision([1,2], [1, 4, 5])

0.3333333333333333

In [22]:
def recall(actual, predicted):
    """Calculate recall

    Parameters
    ----------
    actual : list
        [1st_doc_id, 2nd_doc_id, ...]

    predicted : list
        [1st_doc_id, 2nd_doc_id, ...]
    
    Returns
    -------
    float
        Recall score
    """

    # TODO: implement recall
    tmp = tpfnfp(actual,predicted)
    return tmp[0]/(tmp[0]+tmp[1])

recall([1,2], [1, 4, 5])

0.5

In [39]:
def f1_score(actual, predicted):
    """Calculate f1_score

    Parameters
    ----------
    actual : list
        [1st_doc_id, 2nd_doc_id, ...]

    predicted : list
        [1st_doc_id, 2nd_doc_id, ...]
    
    Returns
    -------
    float
        f1_score
    """

    # TODO: implement f1_score
    p=precision(actual,predicted)
    r=recall(actual,predicted)
    
    f1_score = 0.4
    if p+r==0.0:
        return 0.0
    return 2*p*r/(p+r)

f1_score([1,2], [1, 4, 5])

0.4

In [40]:
def map_score(actual, predicted):
    """Calculate map_score

    Parameters
    ----------
    actual : list
        [1st_doc_id, 2nd_doc_id, ...]

    predicted : list
        [1st_doc_id, 2nd_doc_id, ...]
    
    Returns
    -------
    float
        map_score
    """

    # TODO: implement map_score
    map_score = 0.0
    curr = 0
    
    for i in range(len(predicted)):
        x=predicted[i]
        if x in actual:
            curr+=1
            map_score+=curr/(i+1)
    if len(actual)==0:
        return 0.0
    return map_score/len(actual)

map_score([1,2,3,4,5,6], [1, 7,2, 3,4,5,8,9,10,6])

0.7749999999999999

In [41]:
from math import log

def ndcg_score(actual, predicted):
    """Calculate ndcg_score

    Parameters
    ----------
    actual : list
        [1st_doc_id, 2nd_doc_id, ...]

    predicted : list
        [1st_doc_id, 2nd_doc_id, ...]
    
    Returns
    -------
    float
        ndcg_score
    """

    # TODO: implement recall
    ndcg_score = 0.4
    
    dcggt=0.0
    for i in range(min(len(predicted),len(actual))):
        if i==0:
            dcggt+=1.0
        else:
            dcggt+=1.0/log(i+1,2)
    
    dcgf=0
    for i in range(len(predicted)):
        x=predicted[i]
        if x not in actual:
            continue
        if i==0:
            dcgf+=1
        else:
            dcgf+=1/log(i+1,2)
    
    if dcggt==0:
        return 0.0
    return dcgf/dcggt

ndcg_score([1,2], [1, 4, 5])

0.5

In [94]:
import json 

def validate():
    f=open("validation.json",'r')
    tests_list=json.loads(f.read())
    
    
    for i in range(len(tests_list)):
        test=tests_list[i]
        result=search(test["title_query"], test["plot_query"], test["title_weight"], test["max_size"])
        predicted=[x[0] for x in result]
        actual=test['doc_ids']
        print('Test #', i+1)
        print('actual:',actual)
        print('Predicted:',predicted)
        print('Precision:', precision(actual,predicted))
        print('Recall:', recall(actual,predicted))
        print('F1 score:', f1_score(actual,predicted))
        print('Map score:', map_score(actual,predicted))
        print('NDCG score:', ndcg_score(actual,predicted))
        print()
        
    
validate()

Test # 1
actual: [626, 1812, 3297, 111, 74, 4213, 3296, 4507, 2760, 2039]
Predicted: [626, 1812, 111, 3297, 74, 4213, 3296, 4507, 1116, 2996]
Precision: 0.8
Recall: 0.8
F1 score: 0.8000000000000002
Map score: 0.8
NDCG score: 0.8826728487262255

Test # 2
actual: [5180, 3152]
Predicted: [5180, 3152]
Precision: 1.0
Recall: 1.0
F1 score: 1.0
Map score: 1.0
NDCG score: 1.0

Test # 3
actual: [4643, 4471, 5122, 4836, 3661, 3517, 4105]
Predicted: [4643, 5122, 4471, 4836, 3661, 3517, 4105]
Precision: 1.0
Recall: 1.0
F1 score: 1.0
Map score: 1.0
NDCG score: 1.0

Test # 4
actual: [4670, 4971, 5309, 1136, 4920, 5475, 2871, 4160, 4357, 794, 5749, 4754]
Predicted: [4670, 4971, 4377, 4920, 1136, 5514, 5601, 2871, 4160, 5309, 5749, 754]
Precision: 0.6666666666666666
Recall: 0.6666666666666666
F1 score: 0.6666666666666666
Map score: 0.5224116161616162
NDCG score: 0.7161130021054175

Test # 5
actual: [5903]
Predicted: [5903]
Precision: 1.0
Recall: 1.0
F1 score: 1.0
Map score: 1.0
NDCG score: 1.0

Test #

گزارش


Test # 1

Precision: 0.8

Recall: 0.8

F1 score: 0.8000000000000002

Map score: 0.8

NDCG score: 0.8826728487262255


Test # 2

Precision: 1.0

Recall: 1.0

F1 score: 1.0

Map score: 1.0

NDCG score: 1.0


Test # 3

Precision: 1.0

Recall: 1.0

F1 score: 1.0

Map score: 1.0

NDCG score: 1.0


Test # 4

Precision: 0.6666666666666666

Recall: 0.6666666666666666

F1 score: 0.6666666666666666

Map score: 0.5224116161616162

NDCG score: 0.7161130021054175


Test # 5

Precision: 1.0

Recall: 1.0

F1 score: 1.0

Map score: 1.0

NDCG score: 1.0


Test # 6

Precision: 1.0

Recall: 1.0

F1 score: 1.0

Map score: 1.0

NDCG score: 1.0


Test # 7

Precision: 1.0

Recall: 1.0

F1 score: 1.0

Map score: 1.0

NDCG score: 1.0


<div dir="rtl">
<font face="XB Zar" size=4>
    <h1>
    <b>نکات پایانی</b>
    </h1>
    

1.   سیستم را بهینه پیاده‌سازی کنید تا در زمان کمتری بارگذاری و نمایه‌سازی  انجام شود.
2.   تمام قطعه‌ کدهای بالا باید توسط شما تکمیل شود. نوت‌بوک نهایی باید بدون خطا اجرا شود. اگر تمام کدهای استفاده شده در همین فایل نیست،‌ حتما آن‌ها را نیز به همراه نوت‌بوک در کوئرا آپلود کنید.
3.    اسم توابع و نحوه ورودی گرفتن و خروجی دادن آن‌ها را تغییر ندهید. بقیه اجزای توابع صرفا برای شهود بیشتر شما نوشته شده‌اند و نیازی به نگه‌داری آن‌ها نیست.
4.   در صورت امکان استفاده از کد آماده مشخصا این مورد برای بخش مربوطه ذکر شده است. اگر چیزی در این مورد ذکر نشده نمی‌توانید از کد آماده استفاده کنید.
5.    فایل داده‌ها را در کوئرا آپلود نکنید.
6.    فایل زیپ نهایی و فایل نوت‌بوک حتما به فرمت StudentNumber_phase1 نامگذاری شود.


<b>سالم و موفق باشید.</b>

</font>
</div>